### ISSUE TITLE
How to save metadata of a dataset for external use?

### ISSUE THREAD

**Issue by CBJ16** [facebookresearch/detectron2#1969]
I have trained a model in Colab on a custom dataset object detection in Colab for 3 classes. The lab testing ran smoothly with no issues. I saved the model.pth for use on demo.py, which it was able to detect the object classes in the images, however the metadata (labeling) is off, it shows the default model labeling instead of the dataset classes. It seems that the detection is working however the model is mapping the labels of the object detection to the default model labeling classes.

It seems that I have missed a step or two that I should done before running the demo, but I can't seem to point out where or how, so the reason of me asking "how to save metadata" for external use.

Any idea?

**Comment by MathijsNL** [comment]
Hey CBJ16,

After you register your dataset you can get the metadata from it
```
metadata = MetadataCatalog.get("datasetname")
print(metadata.as_dict())
```
This will output something like:
```
{'name': 'datasetname', 'json_file': 'coco.json', 'image_root': 'dataset_root', 'evaluator_type': 'coco', 'thing_classes': ['class1', 'class2', 'class3', 'class4'], 'thing_dataset_id_to_contiguous_id': {1: 0, 2: 1, 3: 2, 4: 3}}
```

You could save the metadata to json and read it in later, or just read in the dataset as you do with the training.

Let me know if this was what you were looking for.

**Comment by CBJ16** [comment]
Hi MathijsNL,

Many thanks for your kind reply. 
This was, as it seems, the one step missing, I did the following to save it to external drive

import json

with open('/content/drive/My Drive/outputs/metadata.json', 'w') as outfile:
   json.dump(my_dataset_train_metadata.as_dict(), outfile)

Now there is a file named metadata.json

The next question (step) how do I use it in with demo.py? seems it is either an option to add it to the command line as an --opt or I must save this in the colab as additional step when saving my model.pth file as I did here

#saving model weights
from detectron2.modeling import build_model
model = build_model(cfg)
DetectionCheckpointer(model).load(os.path.join(cfg.OUTPUT_DIR, "model_final.pth"))
 
checkpointer = DetectionCheckpointer(model, save_dir="/content/drive/My Drive/outputs/")
checkpointer.save("model_999")  # save to output/model_999.pth

All i want to do is save these metadata to be used in demo.py and label the detected objects.

Many thanks again for your technical advice.

**Comment by CBJ16** [comment]
Just wanna follow up about what I meant above

demo.py uses the metadata associated in cfg.DATASETS.TEST which affects visualization effect.

The metadata associated with demo.py in cfg.DATASETS.TEST is not the same metadata that I registered in Colab, So how do I 'inform' demo.py that it needs to update its cfg.DATASETS.TEST to the correct/new metadata?

**Comment by ppwwyyxx** [comment]
As the `demo.py` script says https://github.com/facebookresearch/detectron2/tree/master/demo it is meant for builtin models (more precisely builtin models on builtin datasets).

You can modify it to add custom datasets.

**Comment by CBJ16** [comment]
Ok.

Do I need to modify the demo.py or the predictor.py as I see in the latter file the metadata.

**Comment by hammooooody** [comment]
I am facing the same issue. It might be that I am not loading the metadata properly. Thanks you @MathijsNL for directing me.

I have saved my dataset into a json file `my_metadata.json`. How do I use it in my `demo.py` code?


`my_metadata.json` looks like
```
{
   "name":"my_dataset_train",
   "json_file":"/content/drive/My Drive/train/_annotations.coco.json",
   "image_root":"/content/drive/My Drive/train",
   "evaluator_type":"coco",
   "thing_classes":[
      "Object 1",
      "Object 2",
      "Object 3",
      "Object 4"
   ],
   "thing_dataset_id_to_contiguous_id":{
      "0":0,
      "1":1,
      "2":2,
      "3":3
   }
}
```
I did this in my `demo.py`


```
def setup_cfg(args):
    # load config from file and command-line arguments
    cfg = get_cfg()
    ##ATTEMPTING CUSTOM DATASETS
    cfg.merge_from_file(args.config_file)
    cfg.DATASETS.TEST = ("/content/drive/My Drive/outputs/metadata.json",) # <-- added this line
    cfg.merge_from_list(args.opts)
    # Set score_threshold for builtin models
    cfg.MODEL.RETINANET.SCORE_THRESH_TEST = args.confidence_threshold
    cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = args.confidence_threshold
    cfg.MODEL.PANOPTIC_FPN.COMBINE.INSTANCES_CONFIDENCE_THRESH = args.confidence_threshold
    cfg.freeze()
    return cfg
```


It managed to draw boxes around the objects I trained it for but no labels.
No sure what I am doing wrong. 

Any help is appreciated. Thank you in advance!

**Comment by MathijsNL** [comment]
Don't do this, that should be the name of your dataset as far as I know.
>     cfg.DATASETS.TEST = ("/content/drive/My Drive/outputs/metadata.json",) # <-- added this line


I am not sure which demo you are using, but for the sake of demonstration I will use:
[https://github.com/facebookresearch/detectron2/blob/master/demo/demo.py](https://github.com/facebookresearch/detectron2/blob/master/demo/demo.py)

On line 14 you import the VisualizationDemo
```
from predictor import VisualizationDemo
```

In the [https://github.com/facebookresearch/detectron2/blob/master/demo/predictor.py](https://github.com/facebookresearch/detectron2/blob/master/demo/predictor.py) on line 24-26 there is:

```
        self.metadata = MetadataCatalog.get(
            cfg.DATASETS.TEST[0] if len(cfg.DATASETS.TEST) else "__unused"
        )
```
That is probably the point where you would need to make sure the right metadata gets used. The way you do that is up to you.

**Comment by meetgandhi123** [comment]
Hello @MathijsNL @hammooooody  @CBJ16  , I am trying to save and load my custom dataset which is in coco format. can you please help me with it? 
I am able to successfully train the model but when I try to save it and use it in another google collab notebook I am not able to get the labels. Please help.

Loading the dataset: 
`!curl -L "https://public.roboflow.com/ds/5c41Z0qTQh?key=EEGSpcnIgA" > roboflow.zip; unzip roboflow.zip; rm roboflow.zip`

Registering the model: 
```
from detectron2.data.datasets import register_coco_instances
register_coco_instances("my_dataset_train1", {}, "/content/train/_annotations.coco.json", "/content/train")
register_coco_instances("my_dataset_val1", {}, "/content/valid/_annotations.coco.json", "/content/valid")
register_coco_instances("my_dataset_test1", {}, "/content/test/_annotations.coco.json", "/content/test")
```
For training:
```
#We are importing our own Trainer Module here to use the COCO validation evaluation during training. Otherwise no validation eval occurs.
from detectron2.engine import DefaultTrainer
from detectron2.evaluation import COCOEvaluator
class CocoTrainer(DefaultTrainer):

  @classmethod
  def build_evaluator(cls, cfg, dataset_name, output_folder=None):

    if output_folder is None:
        os.makedirs("coco_eval", exist_ok=True)
        output_folder = "coco_eval"

    return COCOEvaluator(dataset_name, cfg, False, output_folder)
```

```
#from .detectron2.tools.train_net import Trainer
#from detectron2.engine import DefaultTrainer
from detectron2.config import get_cfg
#from detectron2.evaluation.coco_evaluation import COCOEvaluator
import os

cfg = get_cfg()
cfg.merge_from_file(model_zoo.get_config_file("COCO-Detection/faster_rcnn_X_101_32x8d_FPN_3x.yaml"))
cfg.DATASETS.TRAIN = ("my_dataset_train1",)
cfg.DATASETS.TEST = ("my_dataset_val1",)

cfg.DATALOADER.NUM_WORKERS = 4
cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url("COCO-Detection/faster_rcnn_X_101_32x8d_FPN_3x.yaml")  # Let training initialize from model zoo
cfg.SOLVER.IMS_PER_BATCH = 4
cfg.SOLVER.BASE_LR = 0.001

cfg.SOLVER.WARMUP_ITERS = 300
cfg.SOLVER.MAX_ITER = 450 #adjust up if val mAP is still rising, adjust down if overfit
cfg.SOLVER.STEPS = (300, 450)
cfg.SOLVER.GAMMA = 0.05

cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 64
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 6 #your number of classes + 1

cfg.TEST.EVAL_PERIOD = 500

os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
trainer = CocoTrainer(cfg)
trainer.resume_or_load(resume=False)
trainer.train()
```

```
#test evaluation
from detectron2.data import DatasetCatalog, MetadataCatalog, build_detection_test_loader
from detectron2.evaluation import COCOEvaluator, inference_on_dataset

cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth")
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.85
predictor = DefaultPredictor(cfg)
evaluator = COCOEvaluator("my_dataset_test1", cfg, False, output_dir="./output/")
val_loader = build_detection_test_loader(cfg, "my_dataset_test1")
inference_on_dataset(trainer.model, val_loader, evaluator)
```

`%ls ./output/`

coco_instances_results.json                        last_checkpoint
events.out.tfevents.1601532220.ed734fdf3812.102.2  metrics.json
instances_predictions.pth                          model_final.pth

```
cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth")
cfg.DATASETS.TEST = ("my_dataset_test1", )
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.7   # set the testing threshold for this model
predictor = DefaultPredictor(cfg)
test_metadata = MetadataCatalog.get("my_dataset_test1")
```
Testing:

```
from detectron2.utils.visualizer import ColorMode
import glob

for imageName in glob.glob('/content/test/*jpg'):
  im = cv2.imread(imageName)
  outputs = predictor(im)
  v = Visualizer(im[:, :, ::-1],
                metadata=test_metadata, 
                scale=0.8
                 )
  out = v.draw_instance_predictions(outputs["instances"].to("cpu"))
  cv2_imshow(out.get_image()[:, :, ::-1])
```

**Saving the model:**
```
# Save config to file
with open("output/config.yaml", "w") as f:
    f.write(cfg.dump())

import shutil
shutil.make_archive('custom_model_try3_things', 'zip', '/content/output')
%cp "/content/custom_model_try3_things.zip" "/content/drive/My Drive" 
```

LOADING THE MODEL IN COMPLETELY NEW NOTEBOOK.

```
%mkdir output
%cd output
!unzip "/content/drive/My Drive/custom_model_try3_things.zip"
#!unzip "/content/drive/My Drive/balloon_model.zip"
%cd ..
```

```
import os
cfg = get_cfg()
cfg.merge_from_file("/content/output/config.yaml")
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.7   # set the testing threshold for this model
cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth")
#cfg.merge_from_list(["MODEL.WEIGHTS", "weights.pth"]
predictor = DefaultPredictor(cfg)
my_dataset_train_metadata = MetadataCatalog.get("my_dataset_train1")
print(my_dataset_train_metadata)
```

Testing:

```
from detectron2.utils.visualizer import ColorMode
import glob
im = cv2.imread("/00e481ea1a520175_jpg.rf.49a7562e0ada29cc96e9da3321c96b70.jpg")
outputs = predictor(im)
v = Visualizer(im[:, :, ::-1],
              metadata=my_dataset_train_metadata, 
              scale=0.8
                )
out = v.draw_instance_predictions(outputs["instances"].to("cpu"))
cv2_imshow(out.get_image()[:, :, ::-1])
```

Output:

![download](https://user-images.githubusercontent.com/28274170/94804536-99b5a580-0408-11eb-9a44-b9de2aa349b7.png)

**Comment by MathijsNL** [comment]
Can you let us know the output for:
```
print(my_dataset_train_metadata)
```

In both of the notebooks.

**Comment by meetgandhi123** [comment]
@MathijsNL  Thankyou so much for reply:
Output of print(my_dataset_train_metadata) for 

**Training file:**

Metadata(evaluator_type='coco', image_root='/content/test', json_file='/content/test/_annotations.coco.json', name='my_dataset_test1', thing_classes=['vehicles', 'Ambulance', 'Bus', 'Car', 'Motorcycle', 'Truck'], thing_dataset_id_to_contiguous_id={0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5})

**Testing  File (Uploading the trained model):** 

Metadata(name='my_dataset_train1')
